src/hotspot/share/gc/g1/g1CollectionSetChooser.cpp

```diff
@@ -147,20 +144,27 @@ class G1BuildCandidateRegionsTask : public WorkerTask {
       _regions_added(0) { }
 
     bool do_heap_region(HeapRegion* r) {
-      // We will skip any region that's currently used as an old GC
-      // alloc region (we should not consider those for collection
-      // before we fill them up).
-      if (should_add(r) && !G1CollectedHeap::heap()->is_old_gc_alloc_region(r)) {
-        assert(r->rem_set()->is_complete(), "must be %u", r->hrm_index());
+      // Candidates from marking are always old; also keep regions that are already
+      // collection set candidates (some retained regions) in that list.
+      if (!r->is_old() || r->is_collection_set_candidate()) {
+        // Keep remembered sets and everything for these regions.
+        return false;
+      }
+
+      bool should_add =
+        // We will skip any region that's currently used as an old GC
+        // alloc region (we should not consider those for collection
+        // before we fill them up). Otherwise the Old region must satisfy the liveness
+        // condition and have a complete remembered set.
+        !G1CollectedHeap::heap()->is_old_gc_alloc_region(r) &&
+        G1CollectionSetChooser::region_occupancy_low_enough_for_evac(r->live_bytes()) &&
+        r->rem_set()->is_complete();
```

Comment:
Does this work?

```cpp
assert(!r->rem_set()->is_updating(), "inv");
if (!r->rem_set()->is_complete()) {
  return false;
}
if (G1CollectionSetChooser::region_occupancy_low_enough_for_evac(r->live_bytes())) {
  add_region(r);
} else {
  r->rem_set()->clear(true /* only_cardset */);
}
```

Reply:
I do not think so. Old GC alloc regions need special handling either way (unregistering them as such), and the suggested code may keep them as candidates - they may have a (complete) remembered set here (unless you suggest to change the meaning).

Not sure if breaking up the conjunction any further improves the code; the breakout of the earlier `if (is_old() || !...)` avoids the repetition for the remset clearing.

Something like
```
assert(!r->rem_set()->is_updating(), "must be");
if (!r->rem_set()->is_complete()) {
  return false;
}
bool should_add = !r->is_old_gc_alloc_region() || G1CollectionSetChooser::region_occupancy_low_enough_for_evac(r->live_bytes())
if (should_add) {
  add_region(r);
} else {
  r->rem_set()->clear(true /* only_cardset */);
}
```
would work if that is better for you.

Reply:
Yes, this looks better. The comment should make it clear whether `is_old_gc_alloc_region` is required for correctness or not. (I believe it is not.)